### PR TITLE
Add function for clearing empties

### PR DESCRIPTION
**Pull request by somtochiama**
This pull requests adds an admin command that looks for db_versions in the bookkeeping table that are no longer in `crsql_chnages` table and compacts them

### CODE REVIEW

File: crates/corro-admin/src/lib.rs [hunk 1]
```diff
@@ -294,7 +297,7 @@ async fn handle_conn(
                 Command::Ping => send_success(&mut stream).await,
                 Command::Sync(SyncCommand::Generate) => {
                     info_log(&mut stream, "generating sync...").await;
-                    let sync_state = generate_sync(bookie, agent.actor_id()).await;
+                    let sync_state = generate_sync(bookie, agent.clone().actor_id()).await;
```
Comment (sgrif) [inline crates/corro-admin/src/lib.rs:300]: Is this clone actually needed?

File: crates/corro-admin/src/lib.rs [hunk 2]
```diff
@@ -520,6 +523,61 @@ async fn handle_conn(
 
                     send_success(&mut stream).await;
                 }
+                Command::CompactEmpties => {
+                    let actor_ids: Vec<_> = {
+                        let r = bookie.read("admin compact empties").await;
+                        r.keys().copied().collect()
+                    };
```
Comment (sgrif) [inline crates/corro-admin/src/lib.rs:530]:
```suggestion
                    let actor_ids: Vec<_> = 
bookie.read("admin compact empties")
                        .await
                        .into_keys()
                        .collect();
```

**Review by sgrif — APPROVED** [review]
Looks like it probably needs a rustfmt run

File: crates/corro-agent/src/agent/util.rs [hunk 3]
```diff
@@ -1223,3 +1225,92 @@ pub fn check_buffered_meta_to_clear(
...
+
+    // the rest in the chunks
+    {
+        let mut conn = agent.pool().write_low().await?;
+        let tx = conn.immediate_transaction()?;
+        for v in chunk {
+            store_empty_changeset(&tx, actor_id, v)?;
+        }
+        tx.commit()?;
+    }
```
Comment (sgrif) [inline crates/corro-agent/src/agent/util.rs:1301]:
```suggestion
    for chunk in overwritten_versions.chunks(10) {
        let mut conn = agent.pool().write_low().await?;
        let tx = conn.immediate_transaction()?;
        for v in chunk {
            store_empty_changeset(&tx, actor_id, v)?;
        }
        tx.commit()?;
    }
```

Comment (sgrif) [reply crates/corro-agent/src/agent/util.rs:1301]: `Iterator::chunks` comes from the itertools crate. https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.chunks

File: crates/corro-admin/src/lib.rs [hunk 4]
```diff
@@ -520,6 +523,61 @@ async fn handle_conn(
...
+                            .await
+                        }
+                        _ => {
+                            send_error(
+                                &mut stream,
+                                "Failed to compact empties (check node logs for details)",
+                            )
+                            .await
+                        }
+                    }
```
Comment (sgrif) [inline crates/corro-admin/src/lib.rs:579]: It doesn't seem like there's any reason to use channels for communicating with the spawned task.

```suggestion
                    let agent = agent.clone();
                    let started = std::time::Instant::now();
                    let done = tokio::task::spawn(async move {
                        for actor_id in actor_ids {
                            if let Err(e) =
                                util::clear_empty_versions(agent.clone(), actor_id, Some(&tx)).await
                            {
                                return Some(format!("{e}"));
                            }
                        }

                        None
                    });

                    while let Some(msg) = rx.recv().await {
                        info_log(&mut stream, msg).await;
                    }

                    match done.await {
                        Ok(None) => {
                            let elapsed = started.elapsed().as_secs_f64();
                            info_log(&mut stream, format!(
                                    "Finished compacting empty versions!  Took {} seconds ({} minutes)",
                                    elapsed,
                                    elapsed / 60.0
                                ),
                            ).await;
                            send_success(&mut stream).await
                        }
                        Ok(Some(err)) => {
                            send_error(
                                &mut stream,
                                format!("An error occured while compacting empties: {err}"),
                            )
                            .await
                        }
                        _ => {
                            send_error(
                                &mut stream,
                                "Failed to compact empties (check node logs for details)",
                            )
                            .await
                        }
                    }
```

File: crates/corro-agent/src/agent/util.rs [hunk 5]
```diff
@@ -1223,3 +1225,92 @@ pub fn check_buffered_meta_to_clear(
...
+    //     let mut conn = agent.pool().write_low().await?;
+    //     let tx = conn.immediate_transaction()?;
+    //     for v in ch {
+    //         store_empty_changeset(&tx, actor_id, v)?;
+    //     }
+    //     tx.commit()?;
+    //
+    //     // yield a bit at the end
+    //     tokio::task::yield_now().await;
+    // }
```
Comment (sgrif) [inline crates/corro-agent/src/agent/util.rs:1275]: I'm assuming you didn't meant o commit this?

File: crates/corro-agent/src/agent/util.rs (hunk 3 above)
Comment (sgrif) [inline crates/corro-agent/src/agent/util.rs:1301]:
```suggestion
    let mut conn = agent.pool().write_low().await?;
    let mut tx = conn.immediate_transaction()?;
    let mut chunk_count = 0;
    for v in overwritten_versions {
        store_empty_changeset(&tx, actor_id, v)?;
        chunk_count += 1;
        if chunk_count >= 10 {
            tx.commit()?;
            tx = conn.immediate_transaction()?;
            chunk_count = 0;
        }
    }
    tx.commit()?;
```

**Comment by somtochiama** [comment]
superseded by #223

**Pull request closed without merging**